### ISSUE TITLE
Information on engine CPU / JACK usage

### ISSUE THREAD

**Issue by riban-bw** [zynthian/zynthian-issue-tracking#223]
**Is your feature request related to a problem? Please describe.**
When creating snapshots one must add engines then play with each for a prolonged period to find out whether they are likely to trigger xruns. This hinders artistic endeavours.

**Describe the solution you'd like**
It may be advantageous to classify each engine with its CPU / JACK usage so that a user has some information about the likelihood of engines coexisting. It would further be useful to have a mode where _incompatible_ engines are hidden from the New Engine menus, i.e. don't show any engines that would push the combination over 100%.

**Additional context**
Each engine would need to be benchmarked and given a rating, e.g. processing units. Each platform would need to be benchmarked and given a maximum rating it supports. These benchmarks could be created for standard kids and a function added to Zynthian for users to _calibrate_ their device by running benchmarking.

The benchmarking needs to be very specific, e.g. measuring the CPU usage and time taken to process periods of JACK frames.

Ratings should be available for different configurations, e.g. different samplerates.

**Comment by jofemodo** [comment]
I've though about this in the past, but it's not an easy task. Most of times, CPU usage is highly dependent of the preset being used, what makes the problem more complex. Anyway, with the RBPi4 is difficult to find a engine/preset that push the CPU beyond the limits... BTW,  do you have some example of engine/preset combination that doesn't work on rbpi4?

Regards,

**Comment by riban-bw** [comment]
I agree that this is complex and any implementation is likely to be imperfect but still believe it to have benefit. (At a recent Zynth Club the idea was discussed and benefit was seen by many people. Some of my findings were news and of great interest to members of the club.) I think this would require some script to create the data automatically and maybe create a graph of results showing the minimum and maximum values for an engine for its presets, maybe showing a weighting, e.g. 50% of presets use 3.4% of CPU (but more graph data rather than points). Feel free to assign this to me. I do see the benefit to knowing how to combine engines and how they impact each other, e.g. I may not be able to use Pianoteq whilst setBfree is running (but idle) because setBfree uses 15% CPU all the time but I can use Pianoteq whilst Dexed is idle because Dexed uses 1% CPU when idle. (Those are not real figures, just a plausible example of how this may be useful.)

Regarding specific example, I keep getting sidetracked by code so am yet to really use the Zynthian in anger. I was experiencing issues with setBfree + Pianoteq. I think that collating this data will help me to understand the limitations then I can give more useful specific examples. (I want to use it to make sweet, sweet music, not worry if a glitch is going to ruin a recording or live performance. That rather undermines the beauty you have created with this device and the beauty I would like to create using it!)

**Comment by jofemodo** [comment]
I think predicting the resulting CPU load before loading an engine/preset is near impossible and the amount of work to implement such a system, huge. And of course, you have to think how to maintain such a system updated when engine's code is changed, a newer RBPi is released, etc. It's is milles away from trivial...

In the other hand, having a CPU meter all the time could be very useful for detecting "hedge" cases, while being a lot easier to implement. In fact, it's already implemented and you can enable from webconf. I'm sure you remember ...  ;-) Perhaps we could think a way of having the CPU-meter without disabling the audio level-meter...

Regards,

**Comment by jofemodo** [comment]
I've been testing the CPU load with setBfree + Pianoteq on RBPi4 and i can't get XRuns. It works like a charm.
I also tested adding other heavy engines, like ZynaddSubFX and OBX-d, but i still can't get XRuns.
If don't see any sense to spend time trying to solve problems that doesn't exist on the reference platform, that is RBPi4 now.

Regards,

**Comment by riban-bw** [comment]
I still think that benchmarking is of benefit. I have done some benchmarking a few times and agree that it is very challenging to give an absolute and consistent value for each engine. Some are fairly consistent whilst others vary massively. Maybe we could have a range, indicating that the plugin will generally consume resources within the lower and upper limits of the range. Users then have the concept of the best and worst cases for each engine.

This only works if we can do automated benchmarking with useful metrics. I will assign this to myself so that I can hopefully design such a benchmarking mechanism.
